Option to use external Dataflow Service Account

Lakhtenkov-iv opened the pull request:
This PR is adding `create_service_account` variable.
If it set to `false` than module will not create dataflow service account and will use account email provided in `dataflow_worker_service_account` variable.
Check updated Readme for more details.

Lakhtenkov-iv commented:
@rarsan Could you please review the PR?

Lakhtenkov-iv commented:
- [ ] Merge after version `v1.0.0` released (tag is set).

rarsan commented on README.md:
```diff
@@ -69,9 +70,13 @@ At a minimum, you must have the following roles before you deploy the resources
-**Note about Dataflow permissions**: You must also have permission to impersonate Dataflow worker service account in order to attach that service account to Compute Engine VMs which will execute pipeline operations. In case of default worker service account (i.e. your project's Compute Engine default service account), ensure you have `iam.serviceAccounts.actAs` permission over Compute Engine defa ...
+#### Dataflow permissions
+There are 3 options to manage Dataflow permissions:
+    1. Usage of the default `Compute Engine` account. `roles/dataflow.worker` and other required permissions (example: secrets and kms access) should be granted outside of the module. Permissions to resources created as part of that module `log sync` and `PubSub` topics are would be granted to Default Compute Engine service accountby the module. To use this option don't provide `dataflow_worker_s ...
```
In first option (using default Compute Engine SA): `should be granted outside of the module` is not correct. All necessary permissions are in fact granted inside the module. The only exception is the permission to grant impersonation (`iam.serviceAccounts.actAs` permission via role `roles/iam.serviceAccountUser`) over the default Compute Engine SA. More details in "Note" below in README.

Lakhtenkov-iv replied:
I would argue. `roles/dataflow.worker` access to the KMS, secrets and account impersonation now should be done outside of the module. Here is an example [permissions.tf#96](https://github.com/GoogleCloudPlatform/terraform-splunk-log-export/pull/26/files#diff-f25441e3795a92e3d9321847628159e8e27ae6fc82fe4d548cf1ca8f1dfd9b10R96). Only for option `2` role name not empty string and `use_externally_managed_dataflow_sa` will create the binding.
As you mentioned in another comment we can just remove option `#1` and in case user want to use default Compute Engine SA he can pass its name into as in option `#3`

rarsan replied:
To clarify:
- The only permission this module has stopped at providing is the permission to `impersonate` the default Compute Engine SA as I mentioned above. This gives broad permission to the user/account running this Terraform deployment. That was deemed too risky and obscure from security point of view.
- Granting other permissions (dataflow worker, access to HEC token) to default Compute Engine SA has always been the case and we don't want to break that with this change. It's necessary for a simple first-time run. If risk is not acceptable, or for production purposes, then user is recommended to specify a custom service account.

rarsan replied:
Per other thread, it would be safer to throw an error, rather than inferring what the user wants. Might lead to unintended consequences: maybe the user doesn't want to use or even modify permissions for GCE SA.

We can resolve this as soon as other conversation is resolved.

rarsan commented on README.md:
```diff
@@ -25,6 +25,7 @@ These deployment templates are provided as is, without warranty. See [Copyright
 | <a name="input_region"></a> [region](#input_region) | Region to deploy regional-resources into. This must match subnet's region if deploying into existing network (e.g. Shared VPC). See `subnet` parameter below | `string` |
 | <a name="input_splunk_hec_url"></a> [splunk_hec_url](#input_splunk_hec_url) | Splunk HEC URL to write data to. Example: https://[MY_SPLUNK_IP_OR_FQDN]:8088 | `string` |
 | <a name="input_create_network"></a> [create_network](#input_create_network) | Boolean value specifying if a new network needs to be created. | `bool` |
+| <a name="input_create_service_account"></a> [create_service_account](#input_create_service_account) | (Optional) Defines if service account provided by `dataflow_worker_service_account` variable should be created. If not all permissions (except PubSub topics) for if should be binded externally | `bool` |
```
I think the naming `create_service_account` could be confusing. Consider the case:
create_service_account = true (default)
dataflow_worker_service_account = "" (default)
Module will use the default Compute Engine service but won't actually create a service account, yet `create_service_account` is true....

How about this variable name: `use_externally_managed_service_account`? (default false).
I think this more clearly suggest that both SA is pre-created and its permissions are managed outside the module.

Another alternative to skip this variable altogether: can we add a check in TF on existence of service account (provided by `dataflow_worker_service_account`) and conditionally create it if it doesn't exist?

Lakhtenkov-iv replied:
You are right. It sounds much better. I have renamed.

rarsan commented on README.md:
```diff
@@ -69,9 +70,13 @@ At a minimum, you must have the following roles before you deploy the resources
+There are 3 options to manage Dataflow permissions:
+    1. Usage of the default `Compute Engine` account. `roles/dataflow.worker` and other required permissions (example: secrets and kms access) should be granted outside of the module. Permissions to resources created as part of that module `log sync` and `PubSub` topics are would be granted to Default Compute Engine service accountby the module. To use this option don't provide `dataflow_worker_s ...
+    2. Module is managing Service Account for Dataflow. Module is creating Service Account for Dataflow job and binding all required permissions for it. To use this option set `create_service_account` to `true` and provide `dataflow_worker_service_account` name.
+    3. Custom `Compute Engine` account is managed outside of the module. User should take care of permissions to resources that are managed outside of the module (`roles/dataflow.worker` secrets and kms access) other permissions to resources managed by module are binding by the module. To use this option don't provide existing Compute Engine service account email to `dataflow_worker_service_accou ...
```
For option 3, we should explicitly list the permissions that need to be pre-granted by the user before deploy.
Also, in theory, user could still provide existing Compute Engine SA to `dataflow_worker_service_account` and disable creation of custom SA and permissions management (via say `use_externally_managed_service_account`), right?

Lakhtenkov-iv replied:
Added. There are not so much of them.

rarsan replied:
Thx, resoved.

rarsan reviewed (changes requested):
Thanks for the PR! This is a useful addition.
Suggesting changing variable name and behavior for clarity.

Lakhtenkov-iv commented on main.tf:
```diff
@@ -39,9 +39,20 @@ locals {
+
+  # Grant Dataflow worker service account required IAM roles over external resources
+  # **unless** using an externally managed worker service account (Option 1 and 2):
+  grant_service_account_roles = var.use_externally_managed_dataflow_sa == true
```
@rarsan shouldn't it be `!=`?
```suggestion
  grant_service_account_roles = var.use_externally_managed_dataflow_sa != true
```

rarsan replied:
That's right :) should be `== false` which is option 1 (default) and 2.

Lakhtenkov-iv replied:
One more thing that it is not protacting from invalid combination
```
use_externally_managed_dataflow_sa  = true
dataflow_worker_service_account = ""
```
This could be solved like this:
```suggestion
  grant_service_account_roles = (var.use_externally_managed_dataflow_sa != true && var.dataflow_worker_service_account != "")
```

Lakhtenkov-iv replied:
Apart of that I tested all 3 variants and they looks good.

rarsan replied:
Can we add a variable validation? You're right about this invalid combo but there could be other things that might break, and ideally TF plan fails.

Is conditional validation using other variables now supported? Any other alternative to achieve that check?

Lakhtenkov-iv replied:
As for now it is not possible to reference another variable in 'validation' block. So, it is not possible to fail on 'plan'. It is possible to add some validation to locals and fail on apply. I will check this out.

Lakhtenkov-iv replied:
@rarsan I checked the issue, and want to confirm that there is no good options to solve it now. But here are some workarounds:
1. Use `precondition` for some resource. But this is available only from `1.2.0` and I prefer to avoid bumping version so high. That would limit if somebody running older Terraform.
2. Use this [workaround](https://github.com/hashicorp/terraform/issues/25609#issuecomment-1057614400), but that is quite dirty.

I didn't find any other options to do that.
Please let me know how do you want to proceed.

rarsan replied:
Thanks @Lakhtenkov-iv for looking at these options.
I agree with you. I'm ok with the workaround given that it achieves the desired effect (fail fast in case of invalid input combination), and it's an outlier scenario, i.e. obscure error message not expected to occur often. We can add a TODO to replace with a custom condition check like `precondition` whenever TF version is bumped to >= 1.2.0

Lakhtenkov-iv replied:
Added, please review.

rarsan replied:
LGTM. Thanks. Resolving this.

rarsan commented:
I've updated the variable names and descriptions in changelog and variables.tf. The extra details were needed since `dataflow_worker_service_account` can now take the form of either SA name or SA email address.
Please review the regex condition change as well - I wasn't able to test it out just yet. Please do re-generate the README inputs section as well. I believe that's the last bit remaining.

Lakhtenkov-iv commented:
> I've updated the variable names and descriptions in changelog and variables.tf. The extra details were needed since `dataflow_worker_service_account` can now take the form of either SA name or SA email address. Please review the regex condition change as well - I wasn't able to test it out just yet. Please do re-generate the README inputs section as well. I believe that's the last bit remaining.

Fixed small bug in regexp and tested with 3 different options.
It works good. Also re-generated `README.md` file.
